lib/obj.js

```diff
@@ -1,365 +1,447 @@
-                addFace(
-                    result[1], result[1], undefined, undefined,
-                    result[2], result[2], undefined, undefined,
-                    result[3], result[3], undefined, undefined,
-                    result[4], result[4], undefined, undefined
-                );
-            } else if ((result = facePattern2.exec(line)) !== null) {
-                addFace(
-                    result[1], result[2], result[3], undefined,
-                    result[4], result[5], result[6], undefined,
-                    result[7], result[8], result[9], undefined,
-                    result[10], result[11], result[12], undefined
-                );
-            } else if ((result = facePattern3.exec(line)) !== null) {
-                addFace(
-                    result[1], result[2], result[3], result[4],
-                    result[5], result[6], result[7], result[8],
-                    result[9], result[10], result[11], result[12],
-                    result[13], result[14], result[15], result[16]
-                );
-            } else if ((result = facePattern4.exec(line)) !== null) {
-                addFace(
-                    result[1], result[2], undefined, result[3],
-                    result[4], result[5], undefined, result[6],
-                    result[7], result[8], undefined, result[9],
-                    result[10], result[11], undefined, result[12]
-                );
-            } else if (/^usemtl /.test(line)) {
-                var materialName = line.substring(7).trim();
-                useMaterial(materialName);
-            }
+    // Create a default node in case there are no o/g/usemtl lines in the obj
+    addNode();
+
+    // Parse the obj file
+    return readLines(objPath, parseLine)
+        .then(function() {
+            // Unload resources
+            positions = undefined;
+            normals = undefined;
+            uvs = undefined;
+
+            // Load materials and images
+            return finishLoading(nodes, mtlPaths, objPath);
         });
+}
 
-        stream.on('end', function () {
-            resolve({
-                vertexCount: vertexCount,
-                vertexArray: vertexArray,
-                positionMin: positionMin,
-                positionMax: positionMax,
-                hasUVs: hasUVs,
-                hasNormals: hasNormals,
-                materialGroups: materialGroups,
-                materials: materials,
-                images: images
-            });
+function finishLoading(nodes, mtlPaths, objPath) {
+    nodes = cleanNodes(nodes);
+    if (nodes.length === 0) {
+        throw new RuntimeError(objPath + ' does not have any geometry data');
+    }
+    return loadMaterials(mtlPaths, objPath)
+        .then(function(materials) {
+            var imagePaths = getImagePaths(materials);
+            return loadImages(imagePaths, objPath)
+                .then(function(images) {
```

Comment:
The downside here is then `materials` is not in the closure of the final return (without a helper var).